### PR TITLE
SCP 2345: Replace ContractExe with Builtin contracts in plutus-pab

### PR DESCRIPTION
**Pull request by koslambrou**
Replaced `ContractExe` with `Builtin a` in `plutus-pab`. Here's a resume of what's been done:

- Deleted `Plutus.PAB.ContractCLI` and `Plutus.PAB.Effects.Contract.ContractExe`
- Replaced all occurrences of `ContractExe` with `Builtin a`
- Changed the executable name of `plutus-pab` to `plutus-pab-setup` and only provides the `PSGenerator` command, the migrate command and the `defaultConfig` command.
- Added the module `Plutus.PAB.Run` and `Plutus.PAB.Run.hs` which are used to make an `IO ()` action specialized to a specific `Builtin`. This `IO` action basically replaces the `Plutus.PAB.ContractCLI` module.
- Created the executable `plutus-pab-examples` which provides the available example contracts as `Builtin`.
- Removed the old executables in `plutus-pab` which use `Plutus.PAB.ContractCLI`
- Changed the executable `marlowe-pab` to run the `Builtin` contract `Marlowe`, `MarloweFollow` and `WalletCompanion`. Removed the old Marlowe executables using `Plutus.PAB.ContractCLI`.
- Separated the `PSGenerator` in two parts. The first generates general PS modules, and the second generates PS modules related to a specific `Builtin`. The first can be run with `plutus-pab-setup`, but the other can only run with a specific `Builtin`.
- Removed the possibility to store contract executables in `plutus-pab`.
- Modified nix files to remove used of installed contracts.
- Each specific `Builtin` contracts now have to specify their definitions via `HasDefinitions`.
- Each specific `Builtin` contracts now have to specify the list of modules that need to be converted to PS via `HasPSTypes`.

Pre-submit checklist:
- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: plutus-pab/examples/uniswap/Main.hs
Comment (j-mueller) [inline plutus-pab/examples/uniswap/Main.hs:107]: is this still needed?

File: plutus-pab/src/Plutus/PAB/App.hs
Comment (j-mueller) [inline plutus-pab/src/Plutus/PAB/App.hs:76]: Comment explaining what `a` is

File: plutus-pab/src/Plutus/PAB/Core/ContractInstance.hs
Comment (j-mueller) [inline plutus-pab/src/Plutus/PAB/Core/ContractInstance.hs:124]: this could be extracted into a separate function that returns a `ContractInstanceState t`

File: plutus-pab/src/Plutus/PAB/Db/Beam/ContractStore.hs
Comment (j-mueller) [inline plutus-pab/src/Plutus/PAB/Db/Beam/ContractStore.hs:122]: We could use [`typeRep`](https://hackage.haskell.org/package/base-4.10.1.0/docs/Data-Typeable.html#v:typeRep) to get the name of the type for a more user-friendly error message

File: plutus-pab/src/Plutus/PAB/Effects/Contract/Builtin.hs
Comment (j-mueller) [inline plutus-pab/src/Plutus/PAB/Effects/Contract/Builtin.hs:109]: 👍

File: plutus-pab/src/Plutus/PAB/Effects/Contract/Builtin.hs
Comment (j-mueller) [inline plutus-pab/src/Plutus/PAB/Effects/Contract/Builtin.hs:103]: The typeclass needs some haddocks

File: plutus-pab/src/Plutus/PAB/Run.hs
Comment (j-mueller) [inline plutus-pab/src/Plutus/PAB/Run.hs:55]: Needs haddocks

Something like "PAB entry point for a contract type `a`"

File: plutus-pab/src/Plutus/PAB/Run/PSGenerator.hs
Comment (j-mueller) [inline plutus-pab/src/Plutus/PAB/Run/PSGenerator.hs:55]: Haddocks, we could also add a default definition that returns the empty list (I'm assuming that some (most?) clients won't be written in Purescript so they don't need the PS bridge)

**Review by j-mueller — COMMENTED** [review]
This turned into a rather large change (incl. @silky 's work on restoring the contract state) but it's looking quite good now!

**Review by merivale — CHANGES_REQUESTED** [review]
PureScript stuff looks okay. There are a few follow-on changes that'll be required, but I can do those in a subsequent PR. Just don't delete my comments please. :stuck_out_tongue: 

I'm wondering what the best way would be to generate PS types for the builtin contracts used in particular cases - like the `Marlowe` data type in the marlowe-pab case (please rename this to `MarloweContract`). We might just need a separate PSGenerator for that, but I mention it here in case someone has a better idea...

File: plutus-pab/examples/uniswap/Main.hs
Comment (koslambrou) [reply plutus-pab/examples/uniswap/Main.hs:107]: It's used in the `main` of the same file with `Simulator.runSimulationWith handlers $ do ...`

File: plutus-pab/src/Plutus/PAB/App.hs
Comment (koslambrou) [reply plutus-pab/src/Plutus/PAB/App.hs:77]: Is the comment adequate?

File: plutus-pab/src/Plutus/PAB/Effects/Contract/Builtin.hs
Comment (koslambrou) [reply plutus-pab/src/Plutus/PAB/Effects/Contract/Builtin.hs:106]: Added some haddocks. Adequate enough?

File: plutus-pab/src/Plutus/PAB/Run.hs
Comment (koslambrou) [reply plutus-pab/src/Plutus/PAB/Run.hs:56]: Added some comments. Adequate enough?

File: plutus-pab/src/Plutus/PAB/Run/PSGenerator.hs
Comment (koslambrou) [reply plutus-pab/src/Plutus/PAB/Run/PSGenerator.hs]: Done. Added default implement. Haddocks adequate?

File: marlowe-dashboard-client/plutus-pab.yaml
Comment (koslambrou) [reply marlowe-dashboard-client/plutus-pab.yaml]: My fault. I put back those comments.

File: marlowe-dashboard-client/plutus-pab.yaml
Comment (koslambrou) [reply marlowe-dashboard-client/plutus-pab.yaml]: My fault. I put back those comments.

File: marlowe-dashboard-client/src/Capability/Marlowe.purs
Comment (koslambrou) [reply marlowe-dashboard-client/src/Capability/Marlowe.purs:107]: In that case, I'll let you sort this in a subsequent PR. Thanks!

File: marlowe-dashboard-client/default.nix
Comment (koslambrou) [reply marlowe-dashboard-client/default.nix]: You're right, my fault. I changed `generated-purescript` by calling `psgenerator` and passing the config file for `psapigenerator`.

Comment (koslambrou) [reply marlowe-dashboard-client/default.nix]: What's the point of using the config file to specify the contracts that are being used? We specify the available definitions with the `HasDefinitions` typeclass. You had something else in mind?

File: marlowe-dashboard-client/package.json
Comment (koslambrou) [reply marlowe-dashboard-client/package.json]: Removed them. Thanks!

File: marlowe-dashboard-client/src/API/Contract.purs
Comment (koslambrou) [reply marlowe-dashboard-client/src/API/Contract.purs:50]: Thanks :) I'll let you sort it in a subsequent PR.

File: marlowe/pab/MarloweContracts.hs
Comment (koslambrou) [reply marlowe/pab/MarloweContracts.hs]: Done!

File: plutus-pab-client/src/View/Pretty.purs
Comment (koslambrou) [reply plutus-pab-client/src/View/Pretty.purs:37]: Mmm.... It's not really possible to create an instance of Pretty for Builtin, since it doesn't really have a data constructor with the contract.

@j-mueller How do you think this can be done? Do we need to use `ContractDef` to make that happen?

**Comment by koslambrou** [comment]
@merivale I can't see to be able to compile the module `Language.Marlowe.Client` in `marlowe` because it times out. Memory usage goes incredibly high. How you ever gotten that?

**Comment by kwxm** [comment]
> @merivale I can't seem to be able to compile the module `Language.Marlowe.Client` in `marlowe` because it times out. Memory usage goes incredibly high. How you ever gotten that?

I concur.  I tried to compile Marlowe on our benchmarking machine (which has 8 cores and 32G of memory) and it took over 9 minutes to compile this file, with memory usage peaking over 16G (in the latest version of `master`).  I tried on my laptop as well: it quickly locked it up completely and I had to reboot!

File: plutus-pab-client/src/View/Pretty.purs
Comment (j-mueller) [reply plutus-pab-client/src/View/Pretty.purs:37]: I don't think we need a `prettyBuiltin` - the `Builtin` type has no values, so there isn't anything to pretty print. On the Haskell side we use `Builtin` with a type family. For example, the log messages for a `Builtin a` usually have `ContractDef a` as the concrete type. For `MarloweContract` this is just `MarloweContract` - so we'd need a `prettyMarloweContract`, but the instance needs to be written for each `a` individually, we can't provide a general one as part of `plutus-pab-client`.

File: marlowe-dashboard-client/default.nix
Comment (merivale) [reply marlowe-dashboard-client/default.nix]: I was vaguely wondering if it could be used to generate PS types for the `a` in `Builtin a`, but that's probably not possible. Don't worry about it.

**Comment by merivale** [comment]
I haven't experienced that myself, but it might be the problem @palas mentioned he was having the other day. He thought it might just be his branch, but this makes me wonder whether something unfortunate has made it's way onto master... You should speak to him and @nau about this.

**Comment by koslambrou** [comment]
@j-mueller A few changes since your last review. I delete the module `Plutus.PAB.ParseStringifiedJSON` and removed all log ADTs in `Plutus.PAB.Monitoring.PABLogMsg` which weren't used in the pab.

I also added the CLI command `available` which basically outputs the result of `Contracts.getDefinitions`. It basically replaces the old `installed` command which lists installed contracts.

**Comment by j-mueller** [comment]
@koslambrou :

> added the CLI command `available`

Good idea. I'm happy to merge it now, if we can just get the build to finish!

@merivale : I looked into the marlowe client compilation issue a little bit [here](https://github.com/input-output-hk/plutus/pull/3593) - seems like it is caused by the compilation of the on-chain code bits, which we can't do much about.

**Comment by koslambrou** [comment]
@j-mueller All that's left before merging is for @silky to apply your review for his part of the commit.

**Comment by silky** [comment]
@koslambrou cool I'll do those now

**Pull request merged**
